art/attacks/poisoning/feature_collision_attack.py

```diff
@@ -0,0 +1,279 @@
+# MIT License
+#
+# Copyright (C) The Adversarial Robustness Toolbox (ART) Authors 2020
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated
+# documentation files (the "Software"), to deal in the Software without restriction, including without limitation the
+# rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit
+# persons to whom the Software is furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in all copies or substantial portions of the
+# Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
+# WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+# TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+"""
+This module implements clean-label attacks on Neural Networks.
+"""
+from __future__ import absolute_import, division, print_function, unicode_literals
+
+import logging
+
+import numpy as np
+from tqdm import tqdm
+
+from art.attacks.attack import PoisoningAttackWhiteBox
+from art.estimators import BaseEstimator, NeuralNetworkMixin
+
+from functools import reduce
+
+logger = logging.getLogger(__name__)
+
+
+class FeatureCollisionAttack(PoisoningAttackWhiteBox):
+    """
+    Close implementation of Feature Collision Poisoning Attack by Shafahi, Huang, et al 2018.
+    "Poison Frogs! Targeted Clean-Label Poisoning Attacks on Neural Networks"
+
+    | Paper link: https://arxiv.org/pdf/1804.00792.pdf
+    """
+
+    attack_params = PoisoningAttackWhiteBox.attack_params + [
+        "classifier",
+        "target",
+        "feature_layer",
+        "learning_rate",
+        "decay_coeff",
+        "stopping_tol",
+        "num_old_obj",
+        "max_iter",
+        "similarity_coeff",
+        "watermarking",
+    ]
+
+    _estimator_requirements = (BaseEstimator, NeuralNetworkMixin)
+
+    def __init__(self, classifier, target, feature_layer, learning_rate=500 * 255.0, decay_coeff=0.5,
```

Comment:
This seems like a quite large learning rate, is this the intended default value?

Reply:
Yes this is what was used in the source code. This is set initially very high and is halved while iterating

Reply:
Ok, sounds good